File: Client/game-objects/moveable-object.js
import GameObject from "./game-object.js";
import Direction from "../Direction.js";

export default class MoveableObject extends GameObject {
	/**
	 * @param {number} x
	 * @param {number} y
	 */
	constructor(x, y) {
		super(x, y);
	}

	/**
	 * @param {Direction} direction
	 * @returns {{x:0, y:0}}
	 */
	move(direction) {
		switch (direction) {
			case Direction.up:
				this.y--;
				break;

			case Direction.down:
				this.y++;
				break;

			case Direction.left:
				this.x--;
				break;

			case Direction.right:
				this.x++;
				break;

			default:
				break;
		}

		return { x: this.x, y: this.y };
	}
}

File: Server/HuntTheWumpus.API/Domain/GameObjects/MoveableObject.cs
namespace HuntTheWumpus.API.Domain.GameObjects
{
    public abstract class MoveableObject : GameObject
    {
        protected MoveableObject(uint x, uint y) : base(x, y)
        {
        }

        public (uint X, uint Y) Move(Direction direction)
        {
            switch (direction)
            {
                case Direction.Up:
                    Y--;
                    break;

                case Direction.Down:
                    Y++;
                    break;

                case Direction.Left:
                    X--;
                    break;

                case Direction.Right:
                    X++;
                    break;

                default:
                    break;
            }

            return (X, Y);
        }
    }
}
